internal/marketplace/bundles/bundle.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bundles contains a domain-level representation of the bundles
+// created by the mindpak tool, and logic for working with that model
+package bundles
+
+import (
+	"fmt"
+	"io/fs"
+	"strings"
+
+	"github.com/stacklok/minder/internal/engine"
+	v1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+	"github.com/stacklok/minder/pkg/mindpak"
+)
+
+// Bundle encapsulates operations on the mindpak bundles
+type Bundle interface {
+	// GetMetadata returns the bundle information as-is
+	GetMetadata() *mindpak.Metadata
+	// GetProfile takes the name of a profile in the bundle and attempts to
+	// read it from the bundle, parse it and return an instance of the profile
+	// struct
+	GetProfile(string) (*v1.Profile, error)
+	// ForEachRuleType walks each rule type in the bundle, attempts to read
+	// and parse the rule type, and then applies the specific anonymous
+	// function to the rule type
+	ForEachRuleType(func(*v1.RuleType) error) error
+}
+
+type profileSet = map[string]struct{}
+type bundle struct {
+	original *mindpak.Bundle
+	profiles profileSet
+}
+
+// NewBundle creates an instance of the Bundle interface from a mindpak
+// bundle
+func NewBundle(mpBundle *mindpak.Bundle) Bundle {
+	bundleProfiles := mpBundle.Files.Profiles
+	profiles := make(profileSet, len(bundleProfiles))
+	// build a set of profile names for `GetProfile` to ease lookups
+	for _, profile := range bundleProfiles {
+		profiles[profile.Name] = struct{}{}
+	}
+
+	return &bundle{
+		original: mpBundle,
+		profiles: profiles,
+	}
+}
+
+func (b *bundle) GetMetadata() *mindpak.Metadata {
+	return b.original.Manifest.Metadata
+}
+
+func (b *bundle) GetProfile(name string) (*v1.Profile, error) {
+	// ensure name has file extension
+	name = ensureYamlSuffix(name)
+	// validate that profile exists
+	_, ok := b.profiles[name]
```

Comment:
One thing that got me thinking after writing the library was that `name` may be overloaded in the bundle, leading to ambiguity. Here (as in `pkg/mindpak`) we mean the file name, while the profiles and  rules types [have their own name string](https://github.com/stacklok/minder-rules-and-profiles/blob/6f1bb40c4394d30dd2d837da99cfd52a39d20842/profiles/github/dependabot_go.yaml#L5). Should we rename the name in the bundle to filename? Index by its true name?

Reply:
It may indeed be overloaded by the bundle, we may have a single OCI image reference instead of a file. Ideally, the method of reading rule types and profiles should be pluggable.

Reply:
From an API viewpoint, I think it makes the most sense to refer to a rule type or profile by its actual name (i.e. the name defined inside the profile or rule type). The filename is just an implementation detail.

The way I see it, there are two possible options here:
1. We ensure when building the bundle that the file name matches the profile/rule type which is stored in the file.
2. We add a separate file name or reference field.

Option 2) seems like the most flexible, but I'm not sure if we need that flexibility at the moment.

Reply:
Let's go for simplicity now. You're right that it's an implementation detail.

Reply:
The only tricky part here is that if we unify name + fielame, then how do we specify namespace rule types and profiles?